src/formats/SMI.cpp

```diff
@@ -441,13 +394,88 @@ static void find_rings(
         auto not_hit = std::find(hit_atoms.begin(), hit_atoms.end(), false);
         auto current_atom = static_cast<size_t>(std::distance(hit_atoms.begin(), not_hit));
-        auto& current_atom_bonds = adj_list[current_atom];
+
+        // mark this atom as processed
         hit_atoms[current_atom] = true;
-        if (current_atom_bonds.empty()) {
+
+        // it has no connections... therefore no rings!
+        if (adj_list[current_atom].empty()) {
             continue;
         }
 
-        find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms, current_atom_bonds[0], current_atom);
+        // Search to find all rings for the component (all connected atoms)
+        // for the 'current' atom.
+        // Not guaranteed to fast, nor efficient, but just to work
+        // Also not guaranteed to find the SSSR, but it will find all rings
+        // needed for a given structure to be writen as SMILES
+        // adj_list The adjacentcy list graph
+        // hit_atoms Atoms already encountered
+        // ring_bonds Bonds already known to ring forming bonds
+        // ring_atoms Map from an atom to the number of rings it forms
+
+        // The atoms to process consist of :
+        // current_atom Atom to be checked in this iteration
+        // previous_atom Atom which was bound to the current atom
+
+        std::stack<std::pair<size_t, size_t>> atoms_to_process;
+        atoms_to_process.push({ current_atom, current_atom });
+
+        while (atoms_to_process.size()) {
+
+            size_t previous_atom;
+
+            std::tie(current_atom, previous_atom) = atoms_to_process.top();
+            atoms_to_process.pop();
+
+            auto& current_atom_bonds = adj_list[current_atom];
+
+            // We go through the neighbors backwards because we are using a
+            // stack instead and want to through them in the forward direction
+            // (think of it like the negative of a negative is a positive)
+            // We cannot use a queue as we want to process all neighbors of an
+            // atom first, then move on to the next atom group (allowing for a
+            // depth first like search).
+            for (auto neighbor_iter = current_atom_bonds.rbegin();
+                neighbor_iter != current_atom_bonds.rend();
+                ++neighbor_iter) {
+
+                auto neighbor = *neighbor_iter;
+
+                // prevent back tracking
+                if (neighbor == previous_atom) {
+                    continue;
+                }
+
+                // We've seen this neighbor before! Ring found
+                // but only if we have NOT processed the current atom as this
+                // prevents rings from being created from mutliple directions.
+                if (hit_atoms[neighbor] && !hit_atoms[current_atom]) {
+
+                    // Don't process the same ring connection twice
+                    if (ring_bonds.count(Bond(neighbor, current_atom))) {
```

Comment:
`ring_bonds.count(Bond(neighbor, current_atom)) != 0` or `ring_bonds.find(Bond(neighbor, current_atom)) != ring_bonds.end()` would be more explicit. (it is unfortunate that we have to wait for C++20 to get `std::map::contains`)

Reply:
Yeah, even **C++14** would be nice!

Reply:
Yep =/. I think is it worth supporting old-ish CentOS & Debian releases (CentOS 7 & Debian 8), meaning we need to compile the code with gcc 4.8.5/4.9 respectively. Debian reach EOL in june of this year, but CentOS is still supported until 2024, and computing centers are usually very slow to update.

Do you have anything specific from C++14 you would like to be able to use? We can try to find a C++11 compatible implementation, like with span/string_view/optional.

Reply:
Yeah, I totally understand. I needed to support CentOS 5 for the longest time for an old project, it didn't even support C++11! I don't know if Travis supports this as it is based on Ubuntu, but some CIs support the redhat devtoolset package which allows you to build with newer versions of GCC on old versions of CentOS (this is how I can build **Lemon** on older versions of CentOS). Problem is, many computing centers do not have this package anyway, making it a moot point.

Ideally `std::enable_if_t`, `make_tuple`, and `make_unique`.

Reply:
I think we can add all of these in a `utils` header with a C++11 implementation. Do you want to open an issue about this?